pkg/observer/observer.go

```diff
@@ -0,0 +1,238 @@
+)
+
+// getBlockPublisher returns block publisher for channel
+var getBlockPublisher = func(channelID string) publisher {
+	return peer.BlockPublisher.ForChannel(channelID)
+}
+
+// publisher allows clients to add handlers for various block events
+type publisher interface {
+	// AddWriteHandler adds a handler for KV writes
+	AddWriteHandler(handler gossipapi.WriteHandler)
+}
+
+type dcasClient interface {
+
+	// Get retrieves the DCAS value
+	Get(ns, coll, key string) ([]byte, error)
+
+	// Put puts the DCAS value and returns the key for the value
+	Put(ns, coll string, value []byte) (string, error)
+}
+
+type cfg interface {
+	GetChannels() []string
+}
+
+// Start starts channel observer routines
+func Start(cfg cfg) error {
+
+	if roles.HasRole(observerRole) {
+
+		logger.Infof("peer is an observer, channels to observe: %s", cfg.GetChannels())
+		for _, channelID := range cfg.GetChannels() {
+			observer := &observer{channelID: channelID}
+			observer.Start()
+		}
+
+	} else {
+		logger.Info("peer is not an observer, nothing to do...")
+	}
+
+	return nil
+}
+
+type observer struct {
+	sync.RWMutex
+
+	channelID string
+	dcas      dcasClient
+}
+
+func (o *observer) Start() {
+
+	logger.Debugf("initialize observing on channel: %s", o.channelID)
+
+	// register to receive Sidetree transactions from blocks
+	n := notifier.New(getBlockPublisher(o.channelID))
+
+	sidetreeTxnChannel := n.RegisterForSidetreeTxn()
+
+	go func(channel string, txnsCh <-chan notifier.SidetreeTxn) {
+
+		for {
+			txn, ok := <-txnsCh
+			if !ok {
+				logger.Warnf("received !ok from channel '%s'", channel)
+				return
+			}
+
+			err := o.processSidetreeTxn(channel, txn)
+			if err != nil {
+				logger.Warnf("Failed to process anchor[%s] on channel[%s]: %s", txn.AnchorAddress, channel, err.Error())
+				continue
+			}
+
+			logger.Debugf("Successfully processed anchor[%s] on channel[%s]", txn.AnchorAddress, channel)
+
+		}
+	}(o.channelID, sidetreeTxnChannel)
+
+}
+
+func (o *observer) getDCASClient() dcasClient {
+
+	if o.dcas != nil {
+		return o.dcas
+	}
+
+	dcas := getDCAS(o.channelID)
+
+	o.Lock()
+	defer o.Unlock()
+
+	o.dcas = dcas
+	return dcas
+}
+
+func (o *observer) processSidetreeTxn(channel string, sidetreeTxn notifier.SidetreeTxn) error {
+
+	logger.Debugf("processing sidetree txn:%+v on channel '%s'", sidetreeTxn, channel)
+
+	content, err := o.getContent(sidetreeTxn.AnchorAddress)
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve content for anchor: channel[%s] sidetreeNs[%s] key[%s]", channel, sidetreeNs, sidetreeTxn.AnchorAddress)
+	}
+
+	logger.Debugf("cas content for anchor[%s]: %s", sidetreeTxn.AnchorAddress, string(content))
+
+	af, err := getAnchorFile(content)
+	if err != nil {
+		return errors.Wrapf(err, "failed to unmarshal anchor[%s]", sidetreeTxn.AnchorAddress)
+	}
+
+	return o.processBatchFile(channel, af.BatchFileHash)
+}
+
+func (o *observer) processBatchFile(channel string, batchFileAddress string) error {
```

Comment:
No need for channel arg since you can get it from observer

Reply:
Removed channel from args